components/brave_wallet/browser/tx_state_manager.cc

```diff
@@ -107,41 +111,72 @@ bool TxStateManager::ValueToTxMeta(const base::Value::Dict& value,
   return true;
 }
 
-TxStateManager::TxStateManager(PrefService* prefs)
-    : prefs_(prefs), weak_factory_(this) {}
+TxStateManager::TxStateManager(PrefService* prefs,
+                               value_store::ValueStoreFrontend* store)
+    : prefs_(prefs), initialized_(false), store_(store), weak_factory_(this) {
+  Initialize();
+}
 
 TxStateManager::~TxStateManager() = default;
 
+void TxStateManager::Initialize() {
+  store_->Get(kStorageTransactions, base::BindOnce(&TxStateManager::OnTxsRead,
+                                                   weak_factory_.GetWeakPtr()));
+}
+
+void TxStateManager::OnTxsRead(absl::optional<base::Value> txs) {
+  if (txs) {
+    txs_ = std::move(txs->GetDict());
+  }
+  initialized_ = true;
+}
+
+void TxStateManager::ScheduleWrite(size_t retry_attempts) {
+  // Schedule a delayed task with maximum retry.
+  if (!initialized_) {
+    if (retry_attempts < kMaxInitRetryAttempts) {
+      base::SequencedTaskRunner::GetCurrentDefault()->PostDelayedTask(
+          FROM_HERE,
+          base::BindOnce(&TxStateManager::ScheduleWrite,
+                         weak_factory_.GetWeakPtr(), retry_attempts + 1),
+          base::Milliseconds(kDelayedInitRetryMsec));
+    } else {
+      LOG(ERROR)
+          << "TxStateManager wait for init maximum retry attempts reached.";
+      // Skip and wait for the next one.
+      return;
+    }
+  }
+  store_->Set(kStorageTransactions, base::Value(txs_.Clone()));
+}
+
 void TxStateManager::AddOrUpdateTx(const TxMeta& meta) {
-  ScopedDictPrefUpdate update(prefs_, kBraveWalletTransactions);
-  base::Value::Dict& dict = update.Get();
   const std::string path =
       base::JoinString({GetTxPrefPathPrefix(meta.chain_id()), meta.id()}, ".");
 
-  bool is_add = dict.FindByDottedPath(path) == nullptr;
-  dict.SetByDottedPath(path, meta.ToValue());
+  bool is_add = txs_.FindByDottedPath(path) == nullptr;
+  txs_.SetByDottedPath(path, meta.ToValue());
+  ScheduleWrite(0);
```

Comment:
Can we make all updating methods (AddOrUpdateTx, DeleteTx, etc) return false if initialized_ is false? 
Otherwise a new transaction created by user and then whole set might get overwritten by OnTxsRead

Reply:
fixed in f253205c8cf8d7ec8b87d2e88d4c355d76e4eb7f